language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala

```diff
@@ -244,20 +242,26 @@ object CodeGen {
     filePlans ++ specialPlans
   }
 
-  type LHSIndexedRecords[+RT] = Map[(Identifier, List[Ref.Name]), Record[RT]]
-
-  private[this] def splitNTDs[RT, VT](recordsAndVariants: Iterable[ScopedDataType.DT[RT, VT]])
-    : (LHSIndexedRecords[RT], List[ScopedDataType[Variant[VT]]]) =
-    partitionEithers(recordsAndVariants map {
-      case sdt @ ScopedDataType(qualName, typeVars, ddt) =>
-        ddt match {
-          case r: Record[RT] => Left(((qualName, typeVars.toList), r))
-          case v: Variant[VT] => Right(sdt copy (dataType = v))
-          case e: Enum =>
-            // FixMe (RH) https://github.com/digital-asset/daml/issues/105
-            throw new NotImplementedError("Enum types not supported")
-        }
-    })(breakOut, breakOut)
+  private[this] def splitNTDs[RT, VT](definitions: Iterable[ScopedDataType.DT[RT, VT]]): (
+      Iterable[ScopedDataType[Record[RT]]],
+      Iterable[ScopedDataType[Variant[VT]]],
+      Iterable[ScopedDataType[Enum]]
+  ) = {
+
+    val records = definitions.collect {
+      case sdt @ ScopedDataType(qualName, typeVars, r: Record[RT]) => sdt copy (dataType = r)
+    }
+
+    val variants = definitions.collect {
+      case sdt @ ScopedDataType(qualName, typeVars, v: Variant[VT]) => sdt copy (dataType = v)
+    }
+
+    val enums = definitions.collect {
```

Comment:
I'd rather use `partitionEithers` twice to preserve the exhaustiveness checking, and continue returning the specific types from `splitNTDs`.

Reply:
fixed